Bottom: 8c43a28aead5a5475a2a29f993d5b21bccafa77d
Top:    e4374e594b9ceaefadbb3ca353751ec2fdaa6889
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-02-21 02:35:30 +0000

Refresh of drop-in-pcb-gl-code-various-me

---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index 12a484e..3ccc684 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -300,9 +300,12 @@ hidgl_draw_line (int cap, double width, int x1, int y1, int x2, int y2, double s
   float deltax, deltay, length;
   float wdx, wdy;
   int circular_caps = 0;
+  int hairline = 0;
 
-  if (width == 0.0)
+  if (width == 0.0) {
+    hairline = 1;
     width = 1.0;
+  }
 
   deltax = x2 - x1;
   deltay = y2 - y1;
@@ -333,8 +336,12 @@ hidgl_draw_line (int cap, double width, int x1, int y1, int x2, int y2, double s
   switch (cap) {
     case Trace_Cap:
     case Round_Cap:
-      circular_caps = 1;
-      break;
+      /* Don't bother capping hairlines */
+      if (hairline) {
+        circular_caps = 1;
+        break;
+      }
+      /* Fall through */
 
     case Square_Cap:
     case Beveled_Cap:
@@ -377,9 +384,12 @@ hidgl_draw_arc (double width, int x, int y, int rx, int ry,
   float angle_incr_rad;
   int slices;
   int i;
+  int hairline = 0;
 
-  if (width == 0.0)
+  if (width == 0.0) {
+    hairline = 1;
     width = 1.0;
+  }
 
   inner_r = rx - width / 2.;
   outer_r = rx + width / 2.;
@@ -423,6 +433,10 @@ hidgl_draw_arc (double width, int x, int y, int rx, int ry,
     last_outer_x = outer_x;  last_outer_y = outer_y;
   }
 
+  /* Don't bother capping hairlines */
+  if (hairline)
+    return;
+
   draw_cap (width, x + rx * -cosf (start_angle_rad),
                    y + rx *  sinf (start_angle_rad),
                    start_angle, scale);
